sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala

```diff
@@ -163,64 +192,142 @@ case class BroadcastHashJoin(
+        s"""
+         |// generate join key
+         |${keyVal.code}
+         |// find matches from HashedRelation
+         |UnsafeRow $matched = $anyNull ? null: (UnsafeRow)$relationTerm.getValue(${keyVal.value});
+         |if ($matched != null) {
+         |  ${buildColumns.map(_.code).mkString("\n")}
+         |  $outputCode
+         |}
+         """.stripMargin
+
+      } else {
+        val matches = ctx.freshName("matches")
+        val bufferType = classOf[CompactBuffer[UnsafeRow]].getName
+        val i = ctx.freshName("i")
+        val size = ctx.freshName("size")
+        s"""
+         |// generate join key
+         |${keyVal.code}
+         |// find matches from HashRelation
+         |$bufferType $matches = $anyNull ? null : ($bufferType) $relationTerm.get(${keyVal.value});
+         |if ($matches != null) {
+         |  int $size = $matches.size();
+         |  for (int $i = 0; $i < $size; $i++) {
+         |    UnsafeRow $matched = (UnsafeRow) $matches.apply($i);
+         |    ${buildColumns.map(_.code).mkString("\n")}
+         |    $outputCode
+         |  }
+         |}
+         """.stripMargin
+      }
 
     } else {
-      val matches = ctx.freshName("matches")
-      val bufferType = classOf[CompactBuffer[UnsafeRow]].getName
-      val i = ctx.freshName("i")
-      val size = ctx.freshName("size")
-      s"""
-         | // generate join key
-         | ${keyVal.code}
-         | // find matches from HashRelation
-         | $bufferType $matches = ${anyNull} ? null :
-         |  ($bufferType) $relationTerm.get(${keyVal.value});
-         | if ($matches != null) {
-         |   int $size = $matches.size();
-         |   for (int $i = 0; $i < $size; $i++) {
-         |     UnsafeRow $matched = (UnsafeRow) $matches.apply($i);
-         |     ${buildColumns.map(_.code).mkString("\n")}
-         |     $outputCode
-         |   }
-         | }
-     """.stripMargin
+      // LeftOuter and RightOuter
+
+      // filter the output via condition
+      val checkCondition = if (condition.isDefined) {
+        ctx.currentVars = resultVars
+        val ev = BindReferences.bindReference(condition.get, this.output).gen(ctx)
+        s"""
+         |boolean $valid = true;
+         |if ($matched != null) {
+         |  ${ev.code}
+         |  $valid = !${ev.isNull} && ${ev.value};
+         |}
+         """.stripMargin
+      } else {
+        s"final boolean $valid = true;"
+      }
+
+      if (broadcastRelation.value.isInstanceOf[UniqueHashedRelation]) {
+        s"""
+         |// generate join key
+         |${keyVal.code}
+         |// find matches from HashedRelation
+         |UnsafeRow $matched = $anyNull ? null: (UnsafeRow)$relationTerm.getValue(${keyVal.value});
+         |${buildColumns.map(_.code).mkString("\n")}
+         |${checkCondition.trim}
+         |if (!$valid) {
+         |  // reset to null
+         |  ${buildColumns.map(v => s"${v.isNull} = true;").mkString("\n")}
```

Comment:
Is this right? The most intuitive thing to generate here IMO is a continue. Does that work?